File: 5 Longest Palindromic Substring/C/solution.c
#include <stdlib.h>
#include <string.h>

#include "solution.h"

int expandAroundCenter(char* s, int left, int right);
int max(int a, int b);

char* longestPalindrome(char* s) {
    size_t len = strlen(s);

    if (len == 0) {
        return "";
    }

    int start = 0;
    int end = 0;

    for (int i = 0; i < len; i++) {
        int odd = expandAroundCenter(s, i, i);
        int even = expandAroundCenter(s, i, i + 1);
        int maxLen = max(odd, even);

        if (maxLen > (end - start)) {
            start = i - (maxLen - 1) / 2;
            end = i + maxLen / 2;
        }
    }

    int resultLen = end - start + 1;
    char* result = (char*) malloc(sizeof(char) * resultLen);
    strncpy(result, s + start, resultLen);
    result[resultLen] = '\0';
    return result;
}

int expandAroundCenter(char* s, int left, int right) {
    while (left >= 0 && right < strlen(s) && s[left] == s[right]) {
        left--;
        right++;
    }
    return right - left - 1;
}

int max(int a, int b) {
    return a >= b ? a : b;
}

File: 5 Longest Palindromic Substring/Python/Solution.py
from typing import Self


class Solution:
    def longestPalindrome(self: Self, s: str) -> str:
        l: int = len(s)
        
        if l == 0:
            return ""
        
        def expandAroundCenter(s: str, left: int, right: int) -> int:
            while left >= 0 and right < l and s[left] == s[right]:
                left -= 1
                right += 1
            return right - left - 1
        
        start: int = 0
        end: int = 0
        
        for i in range(l):
            odd: int = expandAroundCenter(s, i, i)
            even: int = expandAroundCenter(s, i, i + 1)
            maxLen: int = max(odd, even)
            
            if maxLen > (end -  start):
                start = i - (maxLen - 1) // 2
                end = i + maxLen // 2
                
        return s[start : end + 1]
        